writer.go

```diff
@@ -0,0 +1,383 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dictzip
+
+import (
+	"bytes"
+	"compress/flate"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"hash/crc32"
+	"io"
+	"math"
+	"os"
+	"time"
+)
+
+const (
+	DefaultChunkSize = math.MaxUint16
+)
+
+const (
+	NoCompression      = flate.NoCompression
+	BestSpeed          = flate.BestSpeed
+	BestCompression    = flate.BestCompression
+	DefaultCompression = flate.DefaultCompression
+	HuffmanOnly        = flate.HuffmanOnly
+)
+
+// Writer implements [io.WriteCloser] for writing dictzip files. Writer writes
+// chunks to a temporary file during write and copies the resulting data to the
+// final file when [Writer.Close] is called.
+//
+// For this reason, [Writer.Close] must be called in order to write the file
+// correctly.
+type Writer struct {
+	// Header is written to the file when [Writer.Close] is called.
+	Header
+
+	// tmp is the temporary file where chunks will be written.
+	tmp *os.File
+
+	// hasData is true if data has been written to the chunk buffer but hasn't
+	// been finalized and written to tmp. We need this because we can't simply
+	// call z.Flush and check chunkBuf.Len due to the fact that flate.Writer
+	// will write sync markers on every call to Flush even if no data has been
+	// written.
+	hasData bool
+
+	// chunkBuf is the current compressed chunk.
+	chunkBuf *bytes.Buffer
+
+	// z is the compression writer used to write the current compressed chunk
+	// to chunkBuf.
+	z *flate.Writer
```

Comment:
Rename this to `compressor` and update comments referring to "flate writer" or "deflate writer" to refer to the "compressor" instead.